### DISCUSSION TITLE
How to test audio with detox?

### DISCUSSION THREAD

**Discussion by taschik** [wix/Detox#2914] [Q&A]
Hi,

Our app uses a lot of different audio and I'd like to test whether certain audio files are being played using react-native-video. What's the best practice with detox?

Thanks a lot!

**Comment by portyu9** [comment]
Hi @taschik 

Detox cannot listen to the device speaker and acoustically verify that a particular audio file was played.

The better approach is to test this at your application's semantic boundary.

For example, if the application chooses among:

```text
audio-a.mp3
audio-b.mp3
audio-c.mp3
```

expose deterministic application state showing which source is being played, or wrap the media player behind an application service that can be observed in the E2E build.

Then Detox can verify:

```text
user action
 -> expected audio source selected
 -> playback enters expected state
 -> UI reflects playing/paused/completed
```

With a media library such as `react-native-video`, callbacks for load/progress/end/error can update application-visible state that Detox can assert.

So Detox can verify **"the application selected and progressed through playback of file X"**, but it cannot directly verify **"file X was audibly emitted by the speaker."**

That final acoustic verification belongs in a lower-level native/media integration test rather than a Detox UI assertion.

Please mark this answer as accepted if it helped, thank you.
